Sources/SourceKit/clangd/ClangLanguageServer.swift

```diff
@@ -24,22 +24,125 @@ final class ClangLanguageServerShim: ToolchainLanguageServer {
 
-  let buildSystem: BuildSystem
+  private let buildSystem: BuildSystem
 
-  let clang: AbsolutePath?
+  private let clangdPath: AbsolutePath
 
-  /// Creates a language server for the given client using the sourcekitd dylib at the specified path.
-  public init(client: Connection, clangd: Connection, buildSystem: BuildSystem,
-              clang: AbsolutePath?) throws {
-    self.clangd = clangd
-    self.buildSystem = buildSystem
-    self.clang = clang
+  private let clangdOptions: [String]
+
+  private let client: MessageHandler
+
+  private var state: LanguageServerState {
+     didSet {
+       for handler in stateChangeHandlers {
+         handler(oldValue, state)
+       }
+     }
+   }
+
+  private var stateChangeHandlers: [(_ oldState: LanguageServerState, _ newState: LanguageServerState) -> Void] = []
+
+  /// Ask the parent of this language server to re-open all documents in this language server.
+  private let reopenDocuments: (ToolchainLanguageServer) -> Void
+
+  /// The `InitializeRequest` with which `clangd` was originally initialized.
+  /// Stored so we can replay the initialization when clangd crashes.
+  private var initializeRequest: InitializeRequest?
+
+  public init(client: MessageHandler,
+              clangdPath: AbsolutePath,
+              buildSettings: BuildSystem?,
+              clangdOptions: [String],
+              reopenDocuments: @escaping (ToolchainLanguageServer) -> Void
+  ) throws {
+    self.client = client
+    self.buildSystem = buildSettings ?? BuildSystemList()
+    self.clangdPath = clangdPath
+    self.clangdOptions = clangdOptions
+    self.state = .connected
+    self.reopenDocuments = reopenDocuments
+  }
+
+  func initialize() throws {
+    let clientToServer: Pipe = Pipe()
+    let serverToClient: Pipe = Pipe()
+
+    let connection = JSONRPCConnection(
+      protocol: MessageRegistry.lspProtocol,
+      inFD: serverToClient.fileHandleForReading.fileDescriptor,
+      outFD: clientToServer.fileHandleForWriting.fileDescriptor
+    )
+
+    self.clangd = connection
+
+    connection.start(receiveHandler: client)
+
+    let process = Foundation.Process()
+
+    if #available(OSX 10.13, *) {
+      process.executableURL = clangdPath.asURL
+    } else {
+      process.launchPath = clangdPath.pathString
+    }
+
+    process.arguments = [
+      "-compile_args_from=lsp",   // Provide compiler args programmatically.
+      "-background-index=false",  // Disable clangd indexing, we use the build
+      "-index=false",             // system index store instead.
+    ] + clangdOptions
+
+    process.standardOutput = serverToClient
+    process.standardInput = clientToServer
+    process.terminationHandler = { [weak self] process in
+      log("clangd exited: \(process.terminationReason) \(process.terminationStatus)")
+      connection.close()
+      if process.terminationStatus != 0 {
+        if let self = self {
+          self.state = .connectionInterrupted
```

Comment:
Similar synchronization issues to the Swift side.

Reply:
Same comment as on the Swift side: 

As I understood it, queue is used to synchronise the requests and responses from `clangd`. Setting the state does not fall into this category, so I did not synchronise it on queue.

Do you want to extend the definition of queue? I don’t think that state really needs to be synchronised, but it wouldn’t hurt either.

Reply:
> I don’t think that state really needs to be synchronised, but it wouldn’t hurt either.

Why not? It's being mutated and read concurrently.

Reply:
I can’t think of a race condition that could happen right now. But it definitely doesn’t hurt to guard it.